File: control.go
package girth

import (
	"encoding/json"
	"fmt"
	"hash/crc32"
	"io"
	"net"
	"os"
	"path/filepath"
	"time"
)

// The control plane runs over TCP. It negotiates the session, exchanges file
// metadata and an integrity checksum, and tells each side where to send UDP
// traffic. After the handshake, the data plane (UDP) takes over.

// Direction of the bulk transfer relative to the client.
const (
	ModeSend = "send" // client pushes a file to the server
	ModeRecv = "recv" // client pulls a file from the server
)

// hello is the client's opening control message.
type hello struct {
	Version   int    `json:"version"`
	Mode      string `json:"mode"` // ModeSend / ModeRecv
	Name      string `json:"name"` // file name (basename used on the wire)
	Size      int64  `json:"size"` // file size (send mode)
	BlockSize int    `json:"blockSize"`
	RateBps   uint64 `json:"rateBps"` // target/initial injection rate
	MaxBps    uint64 `json:"maxBps"`  // ceiling for adaptive rate
	Adaptive  bool   `json:"adaptive"`
	AlphaBps  uint64 `json:"alphaBps"`
	CRC32C    uint32 `json:"crc32c"` // whole-file CRC32C (send mode)

	// Data-plane encryption negotiation (optional).
	Encrypt bool     `json:"encrypt,omitempty"` // client requests encryption
	Ciphers []string `json:"ciphers,omitempty"` // supported AEAD suites, preferred first
	PubKey  []byte   `json:"pubKey,omitempty"`  // client ephemeral X25519 public key
}

// ack is the server's reply.
type ack struct {
	OK      bool   `json:"ok"`
	Err     string `json:"err,omitempty"`
	UDPPort int    `json:"udpPort"` // server's UDP port
	Session uint32 `json:"session"`
	Size    int64  `json:"size"`   // file size (recv mode: server tells client)
	CRC32C  uint32 `json:"crc32c"` // whole-file CRC32C (recv mode)
	Name    string `json:"name"`

	// Data-plane encryption result (echoed when Encrypt was requested).
	Encrypt bool   `json:"encrypt,omitempty"` // server enabled encryption
	Cipher  string `json:"cipher,omitempty"`  // chosen AEAD suite
	PubKey  []byte `json:"pubKey,omitempty"`  // server ephemeral X25519 public key
}

func writeJSON(c net.Conn, v any) error {
	c.SetWriteDeadline(time.Now().Add(30 * time.Second))
	defer c.SetWriteDeadline(time.Time{})
	b, err := json.Marshal(v)
	if err != nil {
		return err
	}
	var lenbuf [4]byte
	putU32(lenbuf[:], uint32(len(b)))
	if _, err := c.Write(lenbuf[:]); err != nil {
		return err
	}
	_, err = c.Write(b)
	return err
}

func readJSON(c net.Conn, v any) error {
	c.SetReadDeadline(time.Now().Add(120 * time.Second))
	defer c.SetReadDeadline(time.Time{})
	var lenbuf [4]byte
	if _, err := io.ReadFull(c, lenbuf[:]); err != nil {
		return err
	}
	n := getU32(lenbuf[:])
	if n > 1<<20 {
		return fmt.Errorf("control message too large: %d", n)
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(c, b); err != nil {
		return err
	}
	return json.Unmarshal(b, v)
}

// TransferParams collects the user-tunable knobs shared by client and server.
type TransferParams struct {
	BlockSize          int
	RateBps            uint64
	MaxBps             uint64
	Adaptive           bool
	AlphaBps           uint64
	ReadWorkers        int
	FeedbackIntervalUs int
	NetTickIntervalUs  int
	ReportInterval     time.Duration
	Verbose            bool
	Encrypt            bool // client: request data-plane encryption
}

// DefaultParams returns sensible defaults.
func DefaultParams() TransferParams {
	return TransferParams{
		BlockSize:          DefaultBlockSize,
		RateBps:            100_000_000, // 100 Mbps
		MaxBps:             10_000_000_000,
		Adaptive:           false,
		AlphaBps:           30_000_000,
		ReadWorkers:        0, // 0 => auto
		FeedbackIntervalUs: 5000,
		NetTickIntervalUs:  10000,
		ReportInterval:     time.Second,
		Verbose:            false,
	}
}

func (p TransferParams) rateConfig(target uint64) RateConfig {
	mode := RateFixed
	if p.Adaptive {
		mode = RateAdaptive
	}
	return RateConfig{
		Mode:      mode,
		TargetBps: target,
		MaxBps:    p.MaxBps,
		Alpha:     float64(p.AlphaBps),
	}
}

func putU32(b []byte, v uint32) {
	b[0] = byte(v)
	b[1] = byte(v >> 8)
	b[2] = byte(v >> 16)
	b[3] = byte(v >> 24)
}
func getU32(b []byte) uint32 {
	return uint32(b[0]) | uint32(b[1])<<8 | uint32(b[2])<<16 | uint32(b[3])<<24
}

// fileCRC32C computes the whole-file CRC32C for end-to-end integrity checks.
func fileCRC32C(f *os.File) (uint32, error) {
	if _, err := f.Seek(0, io.SeekStart); err != nil {
		return 0, err
	}
	h := crc32.New(crcTable)
	buf := make([]byte, 1<<20)
	for {
		n, err := f.Read(buf)
		if n > 0 {
			h.Write(buf[:n])
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return 0, err
		}
	}
	return h.Sum32(), nil
}

func basename(p string) string { return filepath.Base(p) }

// negotiateCryptoServer completes the server side of the key exchange for a
// hello requesting encryption. It returns the fields to echo in the ack plus
// the derived AEAD. If the client did not request encryption it returns a
// disabled result with a nil box.
func negotiateCryptoServer(h hello, session uint32) (enc bool, cipherName string, pub []byte, box *aeadBox, err error) {
	if !h.Encrypt {
		return false, "", nil, nil, nil
	}
	cipherName = chooseCipher(localCiphers(), h.Ciphers)
	if cipherName == "" {
		return false, "", nil, nil, fmt.Errorf("no common cipher suite")
	}
	priv, pub, err := genKeypair()
	if err != nil {
		return false, "", nil, nil, err
	}
	box, err = deriveAEAD(priv, h.PubKey, session, cipherName)
	if err != nil {
		return false, "", nil, nil, err
	}
	return true, cipherName, pub, box, nil
}

File: src/control.rs
//! Control plane (TCP): session negotiation, file metadata, integrity checksum,
//! and X25519 public-key exchange. Length-prefixed JSON, byte-compatible with
//! the Go implementation (LE u32 length prefix; camelCase fields; `omitempty`;
//! base64 `[]byte`).

use crate::protocol::DEFAULT_BLOCK_SIZE;
use crate::rate::{RateConfig, RateMode};
use serde::{Deserialize, Serialize};
use std::io::{self, Read, Write};
use std::net::TcpStream;
use std::time::Duration;

/// Direction of the bulk transfer relative to the client.
pub const MODE_SEND: &str = "send"; // client pushes a file to the server
pub const MODE_RECV: &str = "recv"; // client pulls a file from the server

fn is_false(b: &bool) -> bool {
    !*b
}

/// base64 (standard, padded) serialization for optional byte fields, matching
/// Go's `encoding/json` `[]byte` handling.
mod b64opt {
    use base64::{engine::general_purpose::STANDARD, Engine};
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S: Serializer>(v: &Option<Vec<u8>>, s: S) -> Result<S::Ok, S::Error> {
        match v {
            Some(b) => s.serialize_str(&STANDARD.encode(b)),
            None => s.serialize_none(),
        }
    }
    pub fn deserialize<'de, D: Deserializer<'de>>(d: D) -> Result<Option<Vec<u8>>, D::Error> {
        let opt = Option::<String>::deserialize(d)?;
        match opt {
            Some(s) => STANDARD
                .decode(s.as_bytes())
                .map(Some)
                .map_err(serde::de::Error::custom),
            None => Ok(None),
        }
    }
}

/// The client's opening control message.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Hello {
    pub version: i64,
    pub mode: String,
    pub name: String,
    pub size: i64,
    pub block_size: i64,
    pub rate_bps: u64,
    pub max_bps: u64,
    pub adaptive: bool,
    pub alpha_bps: u64,
    pub crc32c: u32,

    #[serde(default, skip_serializing_if = "is_false")]
    pub encrypt: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ciphers: Option<Vec<String>>,
    #[serde(default, with = "b64opt", skip_serializing_if = "Option::is_none")]
    pub pub_key: Option<Vec<u8>>,

    /// PSK proof of possession: `HMAC-SHA256(token, transcript)` binding the
    /// request fields and the client's ephemeral public key. Absent when
    /// the client is not authenticating. The token itself is never sent.
    #[serde(default, with = "b64opt", skip_serializing_if = "Option::is_none")]
    pub auth_mac: Option<Vec<u8>>,
}

/// The server's reply.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Ack {
    pub ok: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub err: String,
    pub udp_port: i64,
    pub session: u32,
    pub size: i64,
    pub crc32c: u32,
    pub name: String,

    #[serde(default, skip_serializing_if = "is_false")]
    pub encrypt: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub cipher: String,
    #[serde(default, with = "b64opt", skip_serializing_if = "Option::is_none")]
    pub pub_key: Option<Vec<u8>>,

    /// Set when the server requires a PSK and the client did not present a valid
    /// proof, so the client can distinguish "auth needed" from other failures.
    #[serde(default, skip_serializing_if = "is_false")]
    pub auth_required: bool,
    /// The server's reply MAC, authenticating the server and binding the session
    /// id and the server's ephemeral public key.
    #[serde(default, with = "b64opt", skip_serializing_if = "Option::is_none")]
    pub auth_mac: Option<Vec<u8>>,
}

/// Writes a length-prefixed JSON value to the control connection.
pub fn write_json<T: Serialize>(c: &mut TcpStream, v: &T) -> io::Result<()> {
    c.set_write_timeout(Some(Duration::from_secs(30)))?;
    let b = serde_json::to_vec(v).map_err(io::Error::other)?;
    let len = b.len() as u32;
    c.write_all(&len.to_le_bytes())?;
    c.write_all(&b)?;
    c.flush()
}

/// Reads a length-prefixed JSON value from the control connection.
pub fn read_json<T: for<'de> Deserialize<'de>>(c: &mut TcpStream) -> io::Result<T> {
    c.set_read_timeout(Some(Duration::from_secs(120)))?;
    let mut lenbuf = [0u8; 4];
    c.read_exact(&mut lenbuf)?;
    let n = u32::from_le_bytes(lenbuf) as usize;
    if n > (1 << 20) {
        return Err(io::Error::other(format!(
            "control message too large: {}",
            n
        )));
    }
    let mut b = vec![0u8; n];
    c.read_exact(&mut b)?;
    serde_json::from_slice(&b).map_err(io::Error::other)
}

/// User-tunable knobs shared by client and server.
#[derive(Debug, Clone)]
pub struct TransferParams {
    pub block_size: usize,
    pub rate_bps: u64,
    pub max_bps: u64,
    pub adaptive: bool,
    pub alpha_bps: u64,
    pub read_workers: usize,
    pub feedback_interval_us: u32,
    pub net_tick_interval_us: u32,
    pub report_interval: Duration,
    pub verbose: bool,
    pub encrypt: bool,
}

/// Sensible defaults (matching the Go `DefaultParams`).
pub fn default_params() -> TransferParams {
    TransferParams {
        block_size: DEFAULT_BLOCK_SIZE,
        rate_bps: 100_000_000, // 100 Mbps
        max_bps: 10_000_000_000,
        adaptive: false,
        alpha_bps: 30_000_000,
        read_workers: 0, // 0 => auto
        feedback_interval_us: 5000,
        net_tick_interval_us: 10000,
        report_interval: Duration::from_secs(1),
        verbose: false,
        encrypt: false,
    }
}

impl TransferParams {
    pub fn rate_config(&self, target: u64) -> RateConfig {
        RateConfig {
            mode: if self.adaptive {
                RateMode::Adaptive
            } else {
                RateMode::Fixed
            },
            target_bps: target,
            min_bps: 0,
            max_bps: self.max_bps,
            alpha: self.alpha_bps as f64,
        }
    }
}

pub fn basename(p: &str) -> String {
    std::path::Path::new(p)
        .file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| p.to_string())
}
